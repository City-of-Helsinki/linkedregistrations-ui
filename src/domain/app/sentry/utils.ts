import * as Sentry from '@sentry/browser';

const reportError = ({
  data,
  message,
}: {
  data: { error: Record<string, unknown> } & Record<string, unknown>;
  message: string;
}): string => {
  const { error, ...restData } = data;
  const reportObject = {
    extra: {
      data: {
        ...restData,
        currentUrl: window.location.href,
        timestamp: new Date(),
        userAgent: navigator.userAgent,
        errorAsString: JSON.stringify(error),
      },
      level: 'error',
    },
  };

  return Sentry.captureException(message, reportObject);
};

export { reportError };
