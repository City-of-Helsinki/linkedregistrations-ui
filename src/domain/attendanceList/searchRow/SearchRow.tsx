import classNames from 'classnames';
import { FC } from 'react';

import SearchInput from '../../../common/components/searchInput/SearchInput';

import styles from './searchRow.module.scss';

type SearchRowProps = {
  className?: string;
  countText: string;
  onSearchChange: (value: string) => void;
  onSearchSubmit: (value: string) => void;
  searchInputLabel: string;
  searchValue: string;
};
/**
 * Component to set define layout for search input and search result count text
 */
const SearchRow: FC<SearchRowProps> = ({
  className,
  countText,
  onSearchChange,
  onSearchSubmit,
  searchInputLabel,
  searchValue,
}) => {
  return (
    <div className={classNames(styles.searchRow, className)}>
      <span className={styles.count}>{countText}</span>
      <SearchInput
        className={styles.searchInput}
        label={searchInputLabel}
        hideLabel
        onSubmit={onSearchSubmit}
        onChange={onSearchChange}
        placeholder={searchInputLabel}
        value={searchValue}
      />
    </div>
  );
};

export default SearchRow;
