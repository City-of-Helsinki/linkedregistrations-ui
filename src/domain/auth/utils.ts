/* eslint-disable no-console */
import axios from 'axios';

import { APITokens, ExtendedSession, RefreshTokenResponse } from '../../types';
import { User } from '../user/types';

export const getApiTokensRequest = async ({
  accessToken,
  linkedEventsApiScope,
  url,
}: {
  accessToken: string;
  linkedEventsApiScope: string;
  url: string;
}): Promise<APITokens> => {
  const response = await axios.post(url, undefined, {
    headers: {
      Authorization: `Bearer ${accessToken}`,
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    responseType: 'json',
  });

  const linkedevents = response.data[linkedEventsApiScope];

  return { linkedevents };
};

export const refreshAccessTokenRequest = async ({
  clientId,
  clientSecret,
  refreshToken,
  url,
}: {
  clientId: string;
  clientSecret: string;
  refreshToken: string;
  url: string;
}): Promise<RefreshTokenResponse> => {
  const response = await axios.post(
    url,
    {
      client_id: clientId,
      client_secret: clientSecret,
      grant_type: 'refresh_token',
      refresh_token: refreshToken,
    },
    {
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      responseType: 'json',
    }
  );

  return response.data;
};

export const getUserName = ({
  session,
  user,
}: {
  session: ExtendedSession | null;
  user: User | undefined;
}): string => user?.display_name || session?.user?.email || '';
