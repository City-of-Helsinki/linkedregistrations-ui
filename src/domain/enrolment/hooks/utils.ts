import { TFunction } from 'i18next';

import { LEServerError, ServerErrorItem } from '../../../types';
import parseServerErrorMessage from '../../../utils/parseServerErrorMessage';
import pascalCase from '../../../utils/pascalCase';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ErrorObject = Record<string, any>;
type ErrorType = ErrorObject | ErrorObject[] | string;

export const parseEnrolmentServerErrors = ({
  error,
  t,
}: {
  error: ErrorType;
  t: TFunction;
}): ServerErrorItem[] => {
  // LE returns errors as array when trying to create/edit multiple enrolments in same request.
  // In that case call parseEnrolmentServerErrors recursively to get all single errors
  if (Array.isArray(error)) {
    return (error as ErrorObject[]).reduce(
      (previous: ServerErrorItem[], r: ErrorType) => [
        ...previous,
        ...parseEnrolmentServerErrors({ error: r, t }),
      ],
      []
    );
  }

  return typeof error === 'string'
    ? [{ label: '', message: parseServerErrorMessage({ error: [error], t }) }]
    : Object.entries(error).reduce(
        (previous: ServerErrorItem[], [key, error]) => [
          ...previous,
          ...parseEnrolmentServerError({ error, key }),
        ],
        []
      );

  // Get error item for an single error.
  function parseEnrolmentServerError({
    error,
    key,
  }: {
    error: LEServerError;
    key: string;
  }) {
    switch (key) {
      case 'detail':
      case 'non_field_errors':
        return [{ label: '', message: parseServerErrorMessage({ error, t }) }];
      default:
        return [
          {
            label: parseEnrolmentServerErrorLabel({ key }),
            message: parseServerErrorMessage({ error, t }),
          },
        ];
    }
  }

  // Get correct field name for an error item
  function parseEnrolmentServerErrorLabel({ key }: { key: string }): string {
    return t(`enrolment:label${pascalCase(key)}`);
  }
};
