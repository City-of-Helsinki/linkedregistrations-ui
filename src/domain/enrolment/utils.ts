import { AxiosError } from 'axios';

import formatDate from '../../utils/formatDate';
import stringToDate from '../../utils/stringToDate';
import axiosClient from '../app/axios/axiosClient';
import { Registration } from '../registration/types';
import {
  ENROLMENT_INITIAL_VALUES,
  NOTIFICATIONS,
  NOTIFICATION_TYPE,
} from './constants';
import {
  CreateEnrolmentMutationInput,
  Enrolment,
  EnrolmentFormFields,
} from './types';

export const createEnrolment = async (
  input: CreateEnrolmentMutationInput
): Promise<Enrolment> => {
  try {
    const { data } = await axiosClient.post('/signup/', JSON.stringify(input));
    return data;
  } catch (error) {
    throw Error(JSON.stringify((error as AxiosError).response?.data));
  }
};

export const deleteEnrolment = async (
  cancellationCode: string
): Promise<null> => {
  try {
    const { data } = await axiosClient.delete('/signup/', {
      data: JSON.stringify({ cancellation_code: cancellationCode }),
    });
    return data;
  } catch (error) {
    throw Error(JSON.stringify((error as AxiosError).response?.data));
  }
};

export const getEnrolmentNotificationsCode = (
  notifications: string[]
): NOTIFICATION_TYPE => {
  if (
    notifications.includes(NOTIFICATIONS.EMAIL) &&
    notifications.includes(NOTIFICATIONS.SMS)
  ) {
    return NOTIFICATION_TYPE.SMS_EMAIL;
  } else if (notifications.includes(NOTIFICATIONS.EMAIL)) {
    return NOTIFICATION_TYPE.EMAIL;
  } else if (notifications.includes(NOTIFICATIONS.SMS)) {
    return NOTIFICATION_TYPE.SMS;
  } else {
    return NOTIFICATION_TYPE.NO_NOTIFICATION;
  }
};

export const getEnrolmentNotificationTypes = (
  notifications: string
): NOTIFICATIONS[] => {
  switch (notifications) {
    case NOTIFICATION_TYPE.SMS:
      return [NOTIFICATIONS.SMS];
    case NOTIFICATION_TYPE.EMAIL:
      return [NOTIFICATIONS.EMAIL];
    case NOTIFICATION_TYPE.SMS_EMAIL:
      return [NOTIFICATIONS.EMAIL, NOTIFICATIONS.SMS];
    default:
      return [];
  }
};

export const getEnrolmentPayload = (
  formValues: EnrolmentFormFields,
  registration: Registration
): CreateEnrolmentMutationInput => {
  const {
    city,
    dateOfBirth,
    email,
    extraInfo,
    membershipNumber,
    name,
    nativeLanguage,
    notifications,
    phoneNumber,
    serviceLanguage,
    streetAddress,
    zip,
  } = formValues;

  return {
    city: city || null,
    date_of_birth: dateOfBirth
      ? formatDate(stringToDate(dateOfBirth), 'yyyy-MM-dd')
      : null,
    email: email || null,
    extra_info: extraInfo,
    membership_number: membershipNumber,
    name: name || null,
    native_language: nativeLanguage || null,
    notifications: getEnrolmentNotificationsCode(notifications),
    phone_number: phoneNumber || null,
    registration: registration.id as string,
    service_language: serviceLanguage || null,
    street_address: streetAddress || null,
    zipcode: zip || null,
  };
};

export const getEnrolmentDefaultInitialValues = (
  registration: Registration
): EnrolmentFormFields => ({
  ...ENROLMENT_INITIAL_VALUES,
  audienceMaxAge: registration.audience_max_age ?? null,
  audienceMinAge: registration.audience_min_age ?? null,
});

export const getEnrolmentInitialValues = (
  enrolment: Enrolment,
  registration: Registration
): EnrolmentFormFields => {
  return {
    ...getEnrolmentDefaultInitialValues(registration),
    accepted: true,
    city: enrolment.city ?? '-',
    dateOfBirth: enrolment.date_of_birth
      ? formatDate(new Date(enrolment.date_of_birth))
      : '',
    email: enrolment.email ?? '-',
    extraInfo: enrolment.extra_info ?? '-',
    membershipNumber: enrolment.membership_number ?? '-',
    name: enrolment.name ?? '-',
    nativeLanguage: enrolment.native_language ?? '',
    notifications: getEnrolmentNotificationTypes(
      enrolment.notifications as string
    ),
    phoneNumber: enrolment.phone_number ?? '-',
    serviceLanguage: enrolment.service_language ?? '',
    streetAddress: enrolment.street_address ?? '-',
    zip: enrolment.zipcode ?? '-',
  };
};
