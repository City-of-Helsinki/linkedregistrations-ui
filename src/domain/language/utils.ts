import { AxiosError } from 'axios';
import capitalize from 'lodash/capitalize';
import get from 'lodash/get';

import { OptionType, Language, ExtendedSession } from '../../types';
import getLocalisedString from '../../utils/getLocalisedString';
import queryBuilder from '../../utils/queryBuilder';
import { callGet } from '../app/axios/axiosClient';

import {
  LanguagesQueryVariables,
  LanguagesResponse,
  LELanguage,
} from './types';

export const fetchLanguages = async (
  args: LanguagesQueryVariables,
  session: ExtendedSession | null
): Promise<LanguagesResponse> => {
  try {
    const { data } = await callGet({
      session,
      url: languagesPathBuilder(args),
    });
    return data;
  } catch (error) {
    /* istanbul ignore next */
    throw Error(JSON.stringify((error as AxiosError).response?.data));
  }
};

export const languagesPathBuilder = (args: LanguagesQueryVariables): string => {
  const { serviceLanguage } = args;
  const variableToKeyItems = [
    { key: 'service_language', value: serviceLanguage },
  ];

  const query = queryBuilder(variableToKeyItems);

  return `/language/${query}`;
};

export const getLanguageOption = (
  language: LELanguage,
  locale: Language
): OptionType => ({
  label: capitalize(getLocalisedString(language.name, locale)),
  value: language.id,
});

export const sortLanguageOptions = (a: OptionType, b: OptionType): number => {
  const languagePriorities = {
    fi: 3,
    sv: 2,
    en: 1,
  };
  const aPriority = get(languagePriorities, a.value, 0);
  const bPriority = get(languagePriorities, b.value, 0);

  if (aPriority !== bPriority) {
    return bPriority - aPriority;
  }
  return a.label.localeCompare(b.label, undefined, { sensitivity: 'base' });
};
