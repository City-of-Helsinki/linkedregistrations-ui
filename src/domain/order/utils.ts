import { AxiosError } from 'axios';

import getPublicRuntimeConfig from '../../utils/getPublicRuntimeConfig';
import { callGet } from '../app/axios/axiosClient';

import { WebStoreOrder, WebStoreOrderQueryVariables } from './types';

export const fetchWebStoreOrder = async (
  args: WebStoreOrderQueryVariables
): Promise<WebStoreOrder> => {
  const { webStoreApiBaseUrl } = getPublicRuntimeConfig();

  try {
    const { data } = await callGet({
      config: {
        baseURL: webStoreApiBaseUrl,
        headers: { user: args.user },
      },
      session: null,
      url: webStoreOrderPathBuilder(args),
    });
    return data;
  } catch (error) {
    throw Error(JSON.stringify((error as AxiosError).response?.data));
  }
};

export const webStoreOrderPathBuilder = (
  args: WebStoreOrderQueryVariables
): string => {
  const { id } = args;
  return `/order/${id}/`;
};
