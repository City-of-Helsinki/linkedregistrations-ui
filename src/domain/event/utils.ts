import { DateArray, DateTime, EventAttributes, createEvent } from 'ics';
import { TFunction } from 'next-i18next';

import { AddNotificationFn } from '../../common/components/notificationsContext/NotificationsContext';
import { Language } from '../../types';
import getLocalisedString from '../../utils/getLocalisedString';
import { getPlaceFields } from '../place/utils';

import { Event, EventFields } from './types';

export const getEventFields = (event: Event, locale: Language): EventFields => {
  return {
    audienceMaxAge: event.audience_max_age || null,
    audienceMinAge: event.audience_min_age || null,
    description: getLocalisedString(event.description, locale),
    endTime: event.end_time ? new Date(event.end_time) : null,
    freeEvent: !!event.offers[0]?.is_free,
    imageUrl: event.images.find((image) => image?.url)?.url || null,
    keywords: event.keywords.filter((id) => id),
    name: getLocalisedString(event.name, locale),
    offers: event.offers.filter((offer) => !!offer && !offer?.is_free),
    shortDescription: getLocalisedString(event.short_description, locale),
    startTime: event.start_time ? new Date(event.start_time) : null,
  };
};

export const getEventLocationText = ({
  event,
  locale,
}: {
  event: Event;
  locale: Language;
}): string => {
  const { location } = event;
  const {
    addressLocality,
    name: locationName,
    streetAddress,
  } = location
    ? getPlaceFields(location, locale)
    : { addressLocality: '', name: '', streetAddress: '' };

  return (
    [locationName, streetAddress, addressLocality]
      .filter((e) => e)
      .join(', ') || '-'
  );
};

const getDateArray = (date: Date): DateArray => [
  date.getFullYear(),
  date.getMonth() + 1,
  date.getDate(),
  date.getHours(),
  date.getMinutes(),
];

export const getEventAttributes = ({
  event,
  locale,
}: {
  event: Event;
  locale: Language;
}) => {
  const { name, shortDescription } = getEventFields(event, locale);
  const startTime = event.start_time ? new Date(event.start_time) : new Date();
  const endTime = event.end_time ? new Date(event.end_time) : null;

  const start: DateTime = getDateArray(startTime);
  const end: DateTime = endTime ? getDateArray(endTime) : start;
  const eventAttributes: EventAttributes = {
    start,
    end,
    title: name,
    description: shortDescription,
    location: getEventLocationText({ event, locale }),
  };

  return eventAttributes;
};

export const downloadEventIcsFile = async ({
  addNotification,
  event,
  locale,
  t,
}: {
  addNotification: AddNotificationFn;
  event: Event;
  locale: Language;
  t: TFunction;
}) => {
  try {
    const filename = `event_${event.id}.ics`;
    const file: File = await new Promise((resolve, reject) => {
      createEvent(getEventAttributes({ event, locale }), (error, value) => {
        if (error) {
          reject(error);
        }
        resolve(new File([value], filename, { type: 'text/calendar' }));
      });
    });

    const href = URL.createObjectURL(file);
    const anchor = document.createElement('a');
    anchor.download = filename;
    anchor.href = href;
    anchor.click();

    URL.revokeObjectURL(href);
  } catch {
    addNotification({
      label: t('common:eventCalendarButton.failedToCreateFile'),
      type: 'error',
    });
  }
};
