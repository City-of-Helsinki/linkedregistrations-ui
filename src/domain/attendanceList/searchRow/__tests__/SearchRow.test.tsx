import React from 'react';

import { configure, render } from '../../../../utils/testUtils';
import SearchRow, { SearchRowProps } from '../SearchRow';

configure({ defaultHidden: true });

const defaultProps: SearchRowProps = {
  countText: '100 items',
  onSearchChange: jest.fn(),
  onSearchSubmit: jest.fn(),
  searchInputLabel: 'Search',
  searchValue: '',
};

const renderComponent = () => render(<SearchRow {...defaultProps} />);

test('should render component', () => {
  const { container } = renderComponent();

  expect(container).toMatchSnapshot();
});
