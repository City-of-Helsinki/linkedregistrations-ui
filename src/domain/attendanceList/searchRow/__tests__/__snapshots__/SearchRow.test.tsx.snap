// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component 1`] = `
<div>
  <div
    class="searchRow"
  >
    <span
      class="count"
    >
      100 items
    </span>
    <div
      class="TextInput-module_root__2CMNr text-input_hds-text-input__2LODq searchInput hideLabel"
    >
      <label
        class="FieldLabel-module_label__1zrXK "
        for=":r0:"
      >
        Search
      </label>
      <div
        class="TextInput-module_inputWrapper__3Rvel text-input_hds-text-input__input-wrapper__1OqYG"
      >
        <input
          class="TextInput-module_input__1BlHi text-input_hds-text-input__input__GJm5C TextInput-module_hasButton__2KCM1 TextInput-module_hasClearButton__3-tBe text-input_hds-text-input__input-clear__17qr1"
          id=":r0:"
          placeholder="Search"
          type="text"
          value=""
        />
        <div
          class="TextInput-module_buttonWrapper___filA text-input_hds-text-input__buttons__1RMzT"
        >
          <button
            aria-label="TyhjennÃ¤"
            class="TextInput-module_button__1ySMX text-input_hds-text-input__button__1Fh0I TextInput-module_clearButton__bfCLI text-input_hds-text-input__button-clear__2ED7z"
            type="button"
          >
            <svg
              aria-hidden="true"
              aria-label="cross-circle"
              class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
              role="img"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M12 2C17.5228 2 22 6.47715 22 12C22 17.5228 17.5228 22 12 22C6.47715 22 2 17.5228 2 12C2 6.47715 6.47715 2 12 2ZM12 4C7.58172 4 4 7.58172 4 12C4 16.4183 7.58172 20 12 20C16.4183 20 20 16.4183 20 12C20 7.58172 16.4183 4 12 4ZM15 7.5L16.5 9L13.5 12L16.5 15L15 16.5L12 13.5L9 16.5L7.5 15L10.5 12L7.5 9L9 7.5L12 10.5L15 7.5Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <button
            aria-label="Etsi"
            class="TextInput-module_button__1ySMX text-input_hds-text-input__button__1Fh0I"
            type="button"
          >
            <svg
              aria-hidden="true"
              aria-label="search"
              class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
              role="img"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M15 1C19.4183 1 23 4.58172 23 9C23 13.4183 19.4183 17 15 17C13.2001 17 11.539 16.4056 10.2023 15.4023L3.80101 21.8021L2.21002 20.2111L8.60822 13.8117C7.59861 12.4726 7.00002 10.8062 7.00002 9C7.00002 4.58172 10.5817 1 15 1ZM15 3C11.6863 3 9.00002 5.68629 9.00002 9C9.00002 12.3137 11.6863 15 15 15C18.3137 15 21 12.3137 21 9C21 5.68629 18.3137 3 15 3Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
